### PR TITLE
Add support for more delimiters

### PR DESCRIPTION
**Pull request by schimmy**
This handles both #2 and #4. Only question I'd have is the stderr / stdout distinction for the warning message - I think sending to stderr is correct here, though.
I also think we should make a clean break with the past and not support 'comma' etc anymore (which requires a major version bump). I think few enough people use it and it's simple enough that it's not a tough change to update to the new interface.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -28,10 +28,9 @@ endif
 ifeq ($(COVERAGE),1)
 	@go test -cover -coverprofile=$(GOPATH)/src/$@/c.out $@ -test.v
 	@go tool cover -html=$(GOPATH)/src/$@/c.out
-else
-	@echo "TESTING..."
-	@go test $@ -test.v
 endif
+	@echo "TESTING... $@"
+	@go test $@ -test.v
```
Comment (azylman) [inline Makefile]: Why this? This is going to make us run `go test` twice if we're tracking test coverage

File: README.md [hunk 2]
```diff
@@ -23,11 +23,14 @@ go get github.com/Clever/csvlint/cmd/csvlint
 
 ### Options
 
-  * delimiter: the field delimiter to default with
-    * default: comma
-    * valid options: comma, tab
-    * if you want anything else, you're probably doing CSVs wrong
-  * lazyquotes: allow a quote to appear in an unquoted field and a non-doubled quote to appear in a quoted field. _WARNING: your file may pass linting, but not parse in the way you would expect_
+_*WARNING:* to validate strictly against RFC 4180, use the defaults. However we understand that the real world might require a little more flexibility. Note that your file might not parse the way you expect if you do specify other options._
```
Comment (azylman) [inline README.md]: I think this is a bit confusing. I might phrase it as:

```markdown
*NOTE*: The default settings test that a CSV conforms to RFC 4180. By changing the settings, you're no longer strictly adhering to RFC 4180.
```

And I might leave the warning about parsing associated with the `lazyquotes` option

File: cmd/csvlint/main.go [hunk 3]
```diff
@@ -22,16 +24,17 @@ func main() {
...
-	var comma rune
-	switch *delimiter {
-	case "comma":
-		comma = ','
-	case "tab":
-		comma = '\t'
-	default:
-		fmt.Printf("unrecognized delimiter '%s'\n\n", *delimiter)
+	if flag.NFlag() > 0 {
+		fmt.Fprintln(os.Stderr, "Warning: not using defaults, may not validate CSV to RFC 4180")
```
Comment (azylman) [inline cmd/csvlint/main.go]: I like the use of stderr

File: cmd/csvlint/main.go [hunk 4]
```diff
@@ -22,16 +24,17 @@ func main() {
...
+	}
+
+	converted_delimiter, err := strconv.Unquote(`'` + *delimiter + `'`)
```
Comment (azylman) [inline cmd/csvlint/main.go]: Why is this necessary?

File: linter_test.go [hunk 5]
```diff
@@ -15,6 +15,10 @@ var validationTable = []struct {
 	halted   bool
 }{
 	{file: "./test_data/perfect.csv", err: nil, invalids: []CSVError{}},
+	{file: "./test_data/perfect_tab.csv", err: nil, comma: '\t', invalids: []CSVError{}},
+	{file: "./test_data/perfect_pipe.csv", err: nil, comma: '|', invalids: []CSVError{}},
```
Comment (azylman) [inline linter_test.go]: Worth checking the results from some of these?

**Comment by azylman** [comment]
lgtm

File: cmd/csvlint/main.go (hunk 4 above)
Comment (rgarcia) [inline cmd/csvlint/main.go]: also, snake case should fail lint

but it looks like this file isn't linted in test

File: Makefile (hunk 1 above)
Comment (schimmy) [inline Makefile]: hmm, well - it was already doing that, and then not running tests at all if we weren't tracking coverage!
But I did make it conditional: dd39623868f26e5d8e90a1faeb755894d6ebaf81

File: cmd/csvlint/main.go (hunk 4 above)
Comment (schimmy) [inline cmd/csvlint/main.go]: fixing the linter: 4b93977b64d836d3f9adc9b66c3be9d521fe9d88
fixing the underscore: bf74cef0a18e34040d48ecfc8c7dcad0f5da716e
I've noticed that the `gofmt` command has the same problem - it does not recursively lint to `cmd/csvlint/main.go` as advertised. Tried the same wildcard path and it didn't work... we can take a look tomorrow.

Comment (schimmy) [inline cmd/csvlint/main.go]: as for why we need it, it's a little awkward to be able to read in things like `\t':
https://groups.google.com/forum/#!topic/golang-nuts/jYEvC9C3tGo

Without a little bit of hoop-jumping, the input will have to be unicode:
`csvlint --delimiter="    "`
Where those spaces would be an actual tab (it's _really_ annoying to actually paste a tab!!)

Unquoting takes care of this for us. Accepting all code golf, though, it's a bit awkward I think... :-/

File: README.md (hunk 2 above)
Comment (schimmy) [inline README.md]: updated: f47ff49730479ef7f1e8e55114cbe6325dd79a7a

**Comment by schimmy** [comment]
/z assigning back to you to take a look at the golint and gofmt issues - we might need to clean up the rest of our golang makefiles since they are all probably copy-pasted and broken!

File: Makefile [hunk 6]
```diff
@@ -22,14 +22,15 @@ $(PKGS): golint
 	@gofmt -w=true $(GOPATH)/src/$@*/**.go
 ifneq ($(NOLINT),1)
 	@echo "LINTING..."
-	@PATH=$(PATH):$(GOPATH)/bin golint $(GOPATH)/src/$@*/**.go
+	@PATH=$(PATH):$(GOPATH)/bin golint $(GOPATH)/src/$@/...
```
Comment (rgarcia) [inline Makefile]: this will lint all sub-packages as well, which could lead to some redundant output

I'd follow the pattern we have in other Go Makefiles and enumerate the packages we're testing, i.e.

```
SUBPKGS = $(addprefix $(PKG)/, cmd/csvlint)
PKGS = $(PKG) $(SUBPKGS)
```

then the original form of the command here should work

Comment (rgarcia) [inline Makefile]: Also, I think I'm the one guilty of establishing this pattern, but the globs throughout the Makefile should just be `$(GOPATH)/src/$@/*.go`

**Comment by azylman** [comment]
lgtm, barring Raf's comments about redundant linting

**Comment by schimmy** [comment]
not fixing lint in this PR, turned out to be more awkward than I'd like
